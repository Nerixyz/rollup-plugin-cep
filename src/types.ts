import { XMLWriterOptions } from 'xmlbuilder2/lib/interfaces';

export interface Options {
  configPath: string;
  outputPath: string;
  debugOutputPath: string;
  emitDebug: boolean;
  writerOptions?: XMLWriterOptions;
}

export interface Config {
  manifestVersion?: Version;
  bundleId: string;
  bundleName?: string;
  bundleVersion?: Version;
  extensions: Extension[];
  executionEnvironment: ExecutionEnvironment;
  author?: string;
  contact?: { mailto?: string; value?: string };
  legal?: string | { href?: string; value: string };
  abstract?: string | { href?: string; value: string };
}

export interface ExecutionEnvironment {
  hostList: Record<string, Version>;
  localeList?: string[];
  requiredRuntime?: Version;
}

export type Vec2 = [number | string | undefined, number | string | undefined];
export type Version = string | string[] | number | number[];

export interface Extension {
  id: string;
  version?: Version;
  dependencies?: Record<string, string>;
  hostList?: string[];
  dispatchInfos: ExtensionDispatchInfo[];
  debug?: Record<string, number | string>;
}

export interface ExtensionDispatchInfo {
  mainPath?: string;
  scriptPath?: string | { value: string; engine?: string };
  commandLineParameters?: string[];
  autoVisible?: boolean;
  startOn?: string[];
  type?: 'Panel' | 'ModalDialog' | 'Modeless' | 'Custom' | 'Embedded' | 'Dashboard';
  menu?: string | { value: string; placement?: string };
  screenPercentage?: Vec2;
  size?: Vec2;
  maxSize?: Vec2;
  minSize?: Vec2;
  icons?: Record<'Normal' | 'Disabled' | 'RollOver' | 'DarkNormal' | 'DarkRollOver', string>;
  host?: string;
}
