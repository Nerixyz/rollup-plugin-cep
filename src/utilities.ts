import { Config, Vec2, Version } from './types';
import {resolve} from 'path';

export function makeVersion(version: Version): string {
  if (!Array.isArray(version)) return version.toString();
  return version.join('.');
}

export function importConfig(path: string): Config {
  let mod = require(resolve(path));
  if (mod.default) mod = mod.default;
  if (typeof mod === 'function') return mod();
  return mod;
}

export function optAttrToXml<T extends string>(
  obj:
    | undefined
    | string
    | ({
    [x in T]?: string;
  } & { value: string }),
  key: T,
) {
  if (!obj) return undefined;

  return typeof obj === 'string'
    ? obj
    : {
      [`@${key}`]: obj[key],
      '#text': obj.value,
    };
}

export function makeExtensionId(bundleId: string, extensionId: string): string {
  return extensionId.startsWith(bundleId) ? extensionId : `${bundleId}.${extensionId}`;
}

export function objectToXml<K extends string | number, V>(
  obj: Record<K, V> | undefined,
  mapFn: (key: K, value: V) => Record<string, unknown>,
): Record<string, unknown>[] | undefined {
  if (!obj) return undefined;
  return Object.entries(obj).map(([key, value]) => mapFn(key as K, value as V));
}

export function vec2xml(vec?: Vec2) {
  if (!vec) return undefined;
  const [width, height] = vec;
  if (width === undefined && height === undefined) return undefined;

  return {
    Height: height?.toString(),
    Width: width?.toString(),
  };
}
