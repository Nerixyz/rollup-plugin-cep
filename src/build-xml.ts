import { Config, ExecutionEnvironment, ExtensionDispatchInfo } from './types';
import { XMLBuilder } from 'xmlbuilder2/lib/interfaces';
import { create } from 'xmlbuilder2';
import { makeExtensionId, makeVersion, objectToXml, optAttrToXml, vec2xml } from './utilities';

export function buildManifestXml({ extensions, executionEnvironment, ...config }: Config): XMLBuilder {
  const doc = {
    ExtensionManifest: {
      '@Version': makeVersion(config.manifestVersion ?? '9.0'),
      '@ExtensionBundleId': config.bundleId,
      '@ExtensionBundleVersion': makeVersion(config.bundleVersion ?? '1.0.0'),
      '@ExtensionBundleName': config.bundleName,
      Author: config.author,
      Contact: config.contact && { '@mailto': config.contact.mailto, '#text': config.contact.value },
      Legal: optAttrToXml(config.legal, 'href'),
      Abstract: optAttrToXml(config.abstract, 'href'),
      ExtensionList: {
        Extension: extensions.map(ext => ({
          '@Id': makeExtensionId(config.bundleId, ext.id),
          '@Version': ext.version && makeVersion(ext.version),
        })),
      },
      ExecutionEnvironment: executionEnvironmentToXml(executionEnvironment),
      DispatchInfoList: {
        Extension:  extensions.map(extension => ({
          '@Id': makeExtensionId(config.bundleId, extension.id),
          HostList: extension.hostList?.map(host => ({
            Host: {
              '@Name': host,
            },
          })),
          DispatchInfo: extension.dispatchInfos.map(dispatchInfoToXml),
          DependencyList: extension.dependencies && {
            Dependency: objectToXml(extension.dependencies, (id, version) => ({
              '@Id': id,
              '@Version': version,
            })),
          },
        })),
      },
    },
  };

  return create({ encoding: 'UTF-8', standalone: false }, doc);
}

export function buildDebugXml({extensions, bundleId}: Config): XMLBuilder | undefined {
  if(!extensions?.some(ex => ex.debug)) return undefined;

  const doc = {
    ExtensionList: {
      Extension: extensions.filter(ex => ex.debug).map(ex => ({
        '@Id': makeExtensionId(bundleId, ex.id),
        HostList: {
          Host: objectToXml(ex.debug, (name, port) => ({
            '@Name': name,
            '@Port': port,
          })),
        },
      })),
    }
  };

  return create({ encoding: 'UTF-8' }, doc);
}

function executionEnvironmentToXml({ hostList, localeList, requiredRuntime }: ExecutionEnvironment) {
  return {
    HostList: objectToXml(hostList, (name, version) => ({
      Host: {
        '@Name': name,
        '@Version': makeVersion(version),
      },
    })),
    LocaleList: localeList && {
      Locale: localeList.map(loc => ({
        '@Code': loc,
      })),
    },
    RequiredRuntimeList: requiredRuntime && {
      RequiredRuntime: {
        '@Name': 'CSXS',
        '@Version': makeVersion(requiredRuntime),
      },
    },
  };
}

function dispatchInfoToXml({ commandLineParameters, ...info }: ExtensionDispatchInfo) {
  return {
    '@Host': info.host,
    Resources: (info.mainPath || info.scriptPath || commandLineParameters) && {
      MainPath: info.mainPath,
      ScriptPath: optAttrToXml(info.scriptPath, 'engine'),
      CEFCommandLine: commandLineParameters && {
        Parameter: commandLineParameters,
      },
    },
    Lifecycle: {
      AutoVisible: (info.autoVisible ?? true)?.toString(),
      StartOn: info.startOn && {
        Event: info.startOn,
      },
    },
    UI: (info.type ||
      info.menu ||
      info.screenPercentage ||
      info.size ||
      info.maxSize ||
      info.minSize ||
      info.icons) && {
      Type: info.type,
      Menu: optAttrToXml(info.menu, 'placement'),
      Geometry: (info.screenPercentage || info.size || info.minSize || info.maxSize) && {
        ScreenPercentage: vec2xml(info.screenPercentage),
        Size: vec2xml(info.size),
        MaxSize: vec2xml(info.maxSize),
        MinSize: vec2xml(info.minSize),
      },
      Icons: info.icons && {
        Icon: objectToXml(info.icons, (key, value) => ({
          '@Type': key,
          '#text': value,
        })),
      },
    },
  };
}
