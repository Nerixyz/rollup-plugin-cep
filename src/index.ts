import { create } from 'xmlbuilder2';
import { Plugin } from 'rollup';
import { Options } from './types';
import { buildDebugXml, buildManifestXml } from './build-xml';
import { importConfig } from './utilities';

export default function cep(input?: Partial<Options>): Plugin {
  const options: Options = {
    configPath: 'cep.config.js',
    outputPath: 'CSXS/manifest.xml',
    debugOutputPath: '.debug',
    emitDebug: true,
    writerOptions: {
      prettyPrint: true,
    },
    ...input,
  };

  return {
    name: 'cep',
    buildStart() {
      this.addWatchFile(options.configPath);
    },
    generateBundle() {
      const config = importConfig(options.configPath);
      const built = buildManifestXml(config);
      this.emitFile({
        fileName: options.outputPath,
        type: 'asset',
        source: built.toString(options.writerOptions),
      });
      if(!options.emitDebug) return;

      const debug = buildDebugXml(config);
      if(!debug) return;

      this.emitFile({
        fileName: options.debugOutputPath,
        type: 'asset',
        source: debug.toString(options.writerOptions),
      });
    },
  };
}
