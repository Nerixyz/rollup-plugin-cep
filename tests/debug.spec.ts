// @ts-ignore -- no
import { buildDebugXml } from '../src/build-xml';
import { DefaultConfig, FullConfig } from './test.configs';

const xsd = require('ksys-libxmljs2-xsd');

const schema = xsd.parseFile('tests/debug.xsd');

describe('buildDebugXml', () => {
  it('should generate a valid debug config', () => {
    const builder = buildDebugXml(FullConfig);
    expect(builder).toBeDefined();

    const xml = builder!.toString({prettyPrint: true});

    expect(schema.validate(xml)).toBe(null);
    expect(xml).toMatchSnapshot();
  });
  it('should no emit xml if there\'s no debug info', function() {
    const builder = buildDebugXml(DefaultConfig);
    expect(builder).toBe(undefined);
  });
})
