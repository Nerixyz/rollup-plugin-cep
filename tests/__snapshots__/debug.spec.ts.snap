// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildDebugXml should generate a valid debug config 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<ExtensionList>
  <Extension Id=\\"de.nerixyz.bundle.extension1\\">
    <HostList>
      <Host Name=\\"AEFT\\" Port=\\"1337\\"/>
    </HostList>
  </Extension>
  <Extension Id=\\"de.nerixyz.bundle.extension2\\">
    <HostList>
      <Host Name=\\"AEFT\\" Port=\\"1337\\"/>
    </HostList>
  </Extension>
</ExtensionList>"
`;
